### ISSUE TITLE
Cilium 1.9.5 + EKS & IMDSv2 seems not working

### ISSUE THREAD

**Issue by elagree** [cilium/cilium#15520]
Hello,

With my team we are working on Cilium (1.9.5) integration on a freshly EKS cluster.

We follow the official [doc](https://docs.cilium.io/en/latest/gettingstarted/k8s-install-eks/) and it's working great.

For obvious [reasons](https://docs.aws.amazon.com/eks/latest/userguide/best-practices-security.html) we would like to use only Metadatav2 with our EC2 workers but it seems not working correctly.

```
kube-system   cilium-dlrhr                      0/1     Error               4          112s
kube-system   cilium-h47r4                      0/1     Running             4          112s
kube-system   cilium-node-init-d87d5            1/1     Running             0          112s
kube-system   cilium-node-init-z8hz5            1/1     Running             0          112s
kube-system   cilium-operator-f44f7c878-hk2pd   1/1     Running             0          112s
kube-system   cilium-operator-f44f7c878-rzfc8   1/1     Running             0          112s
kube-system   coredns-6987776bbd-6jn84          0/1     ContainerCreating   0          11m
kube-system   coredns-6987776bbd-dqbdq          0/1     ContainerCreating   0          11m
kube-system   hubble-relay-777ddbdccf-2j8k8     0/1     ContainerCreating   0          112s
kube-system   hubble-ui-58bbbc8b95-2dzgc        0/3     ContainerCreating   0          112s
...
``` 
When checking the logs from Cilium pods I can see it has some difficulties to recover information provided by the metadata :

```
level=warning msg="Unable to mutate nodeResource" error="InstanceID of own EC2 instance is empty" retryCount=7 subsys=nodediscovery
level=debug msg="Missing io.cilium.network.ipv4-cilium-host. Annotation required when IPSec Enabled" k8sNodeID=XXX nodeName=ip-xxx.region.compute.internal subsys=k8s
level=debug msg="Missing io.cilium.network.ipv6-cilium-host. Annotation required when IPSec Enabled" k8sNodeID=XXX nodeName=ip-xxx.region.compute.internal subsys=k8s
level=debug msg="Empty IPv4 CIDR annotation in node" k8sNodeID=XXX nodeName=ip-xxx.region.compute.internal subsys=k8s
level=debug msg="Empty IPv6 CIDR annotation in node" k8sNodeID=XXX nodeName=ip-xxx.region.compute.internal subsys=k8s
level=debug msg="Empty IPv4 health endpoint annotation in node" k8sNodeID=XXX nodeName=ip-xxx.region.compute.internal subsys=k8s
level=debug msg="Empty IPv6 health endpoint annotation in node" k8sNodeID=XXX nodeName=ip-xxx.region.compute.internal subsys=k8s
level=warning msg="Unable to mutate nodeResource" error="InstanceID of own EC2 instance is empty" retryCount=8 subsys=nodediscovery
level=debug msg="Missing io.cilium.network.ipv4-cilium-host. Annotation required when IPSec Enabled" k8sNodeID=XXX nodeName=ip-xxx.region.compute.internal subsys=k8s
level=debug msg="Missing io.cilium.network.ipv6-cilium-host. Annotation required when IPSec Enabled" k8sNodeID=XXX nodeName=ip-xxx.region.compute.internal subsys=k8s
level=debug msg="Empty IPv4 CIDR annotation in node" k8sNodeID=XXX nodeName=ip-xxx.region.compute.internal subsys=k8s
level=debug msg="Empty IPv6 CIDR annotation in node" k8sNodeID=XXX nodeName=ip-xxx.region.compute.internal subsys=k8s
level=debug msg="Empty IPv4 health endpoint annotation in node" k8sNodeID=XXX nodeName=ip-xxx.region.compute.internal subsys=k8s
level=debug msg="Empty IPv6 health endpoint annotation in node" k8sNodeID=XXX nodeName=ip-xxx.region.compute.internal subsys=k8s
level=warning msg="Unable to mutate nodeResource" error="InstanceID of own EC2 instance is empty" retryCount=9 subsys=nodediscovery
level=fatal msg="Could not create or update CiliumNode resource, despite retries" subsys=nodediscovery
```

I saw when debugging (but I'm not able to find it again) that Cilium wasn't able to connect to Metadata and get a 401/Unauthorized like it's trying to get info without having a valid token (basically requesting a token with a PUT to "http://169.254.169.254/latest/api/token" with the header "X-aws-ec2-metadata-token-ttl-seconds: 21600").

Is there any plan to support IMDSv2 ? I think it's now integrated in the AWS SDK.

Thanks,

Regards

**Comment by elagree** [comment]
With the help of my coworker @[Smana](https://github.com/Smana) we had a look on the Cilium code especially the [nodediscovery.go](https://github.com/cilium/cilium/blob/master/pkg/nodediscovery/nodediscovery.go) and it seems that the [metadata.go](https://github.com/cilium/cilium/blob/master/pkg/aws/metadata/metadata.go) used by the node discovery doesn't cover the [IMDSv2](https://aws.amazon.com/fr/blogs/security/defense-in-depth-open-firewalls-reverse-proxies-ssrf-vulnerabilities-ec2-instance-metadata-service/)

So I tried to make use of the official [SDK](https://github.com/aws/aws-sdk-go) and especially this [piece of code](https://aws.github.io/aws-sdk-go-v2/docs/sdk-utilities/ec2-imds/).
Unfortunately the latest stable version of Cilium (v1.9.5) is still using an old version of the SDK (v0.24.0) and it requires a lot of changes in order to upgrade the package I guess:
```console
github.com/cilium/cilium/pkg/aws/endpoints imports
        github.com/aws/aws-sdk-go-v2/aws/endpoints: module github.com/aws/aws-sdk-go-v2@latest found (v1.3.3), but does not contain package github.com/aws/aws-sdk-go-v2/aws/endpoints
github.com/cilium/cilium/pkg/aws/eni imports
        github.com/aws/aws-sdk-go-v2/aws/awserr: module github.com/aws/aws-sdk-go-v2@latest found (v1.3.3), but does not contain package github.com/aws/aws-sdk-go-v2/aws/awserr
github.com/cilium/cilium/pkg/aws/eni/limits imports
        github.com/aws/aws-sdk-go-v2/aws/external: module github.com/aws/aws-sdk-go-v2@latest found (v1.3.3), but does not contain package github.com/aws/aws-sdk-go-v2/aws/external
github.com/cilium/cilium/pkg/ipam/allocator/aws imports
        github.com/aws/aws-sdk-go-v2/aws/ec2metadata: module github.com/aws/aws-sdk-go-v2@latest found (v1.3.3), but does not contain package github.com/aws/aws-sdk-go-v2/aws/ec2metadata
```
Anyway I saw that the `master` branch includes a recent version of the SDK, I'll try to propose a PR, hoping that it will be included in the next stable version (edited)

**Comment by ungureanuvladvictor** [comment]
@elagree -- thank you for the investigation. I reached the same conclusion. Back in the day we rolled out our own ec2 metadata client so we do not vendor the full `aws-go-sdk` in the `cilium-agent` (keeping the memory usage of the agent as small as possible).

I think now with the move to `aws-go-sdk-v2` and the support of go modules we can remove this code and replace it with https://github.com/aws/aws-sdk-go-v2/tree/main/feature/ec2/imds.

Let me know which path you'll take, I can put the PR up if you get caught with too many things.
